File: 0160. Intersection of Two Linked Lists/Solution.java
//one solution 
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA==null || headB==null) return null;
        
        ListNode a = headA;
        ListNode b = headB;
        
        int lenA = lenLL(headA);
        int lenB = lenLL(headB);

        int diff = Math.abs(lenA - lenB);

        if(lenA>lenB){
            int i = 0;
            while(i<diff){
                a = a.next;
                i++;
            }
        }
        else{
            int i = 0;
            while(i<diff){
                b = b.next;
                i++;
            }
        }

        while(a!=null){
            if(a==b){
                return a;
            }
            a = a.next;
            b = b.next;
        }

        return null;
    }
    public int lenLL(ListNode head){
        int i = 0;
        ListNode temp = head;
        while(temp!=null){
            i++;
            temp = temp.next;
        }
        return i;
    }
}

File: 0160. Intersection of Two Linked Lists/Solution.py
# one solution 

class Solution(object):
    def getIntersectionNode(self, headA, headB):

        lenA = self.length(headA)
        lenB = self.length(headB)

        a = headA
        b = headB

        diff = abs(lenA - lenB)

        if lenA > lenB:
            for _ in range(diff):
                a = a.next
        else:
            for _ in range(diff):
                b = b.next

        while a and b:
            if a == b:
                return a
            a = a.next
            b = b.next

        return None

    def length(self, head):
        count = 0
        temp = head
        while temp:
            count += 1
            temp = temp.next
        return count
    

#another slightly better solution 
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        a = headA
        b = headB

        while a!=b:
            a = a.next if a!=None else headB
            b = b.next if b!=None else headA

        return a

        